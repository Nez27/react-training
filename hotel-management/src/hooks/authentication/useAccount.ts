import { useQuery } from '@tanstack/react-query';

// Services
import { getCurrentAccount } from '@service/authenticationService';

/**
 * Get data of account currently login
 * @returns The data of account currently login
 */
const useAccount = () => {
  let isAuthenticated: boolean = false;
  const { data: account, isPending } = useQuery({
    queryKey: ['account'],
    queryFn: getCurrentAccount,
  });

  if (account) {
    isAuthenticated = account.role === 'authenticated';
  }

  return { isPending, account, isAuthenticated };
};

export { useAccount };
