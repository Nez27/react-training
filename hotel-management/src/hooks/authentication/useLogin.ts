import { useMutation, useQueryClient } from '@tanstack/react-query';
import { useNavigate } from 'react-router-dom';
import toast from 'react-hot-toast';

// Services
import { login as loginFn } from '@service/authenticationService';

/**
 * Login web
 * @returns The login function and isPending boolean
 */
const useLogin = () => {
  const queryClient = useQueryClient();
  const navigate = useNavigate();

  const { mutate: login, isPending } = useMutation({
    mutationFn: loginFn,
    onSuccess: (account) => {
      queryClient.setQueryData(['account'], account.user);

      navigate('/');
    },
    onError: (err) => {
      console.error(err.message);
      toast.error('Email or password not correct!');
    },
  });

  return { login, isPending };
};

export { useLogin };
