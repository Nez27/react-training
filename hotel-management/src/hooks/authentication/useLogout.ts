import { useMutation, useQueryClient } from '@tanstack/react-query';
import { useNavigate } from 'react-router-dom';

// Services
import { logout as logoutFn } from '@service/authenticationService';

/**
 * Logout web
 * @returns The logout function and isPending boolean
 */
const useLogout = () => {
  const navigate = useNavigate();
  const queryClient = useQueryClient();

  const { mutate: logout, isPending } = useMutation({
    mutationFn: logoutFn,
    onSuccess: () => {
      queryClient.removeQueries();
      navigate('/login', { replace: true });
    },
  });

  return { logout, isPending };
};

export { useLogout };
