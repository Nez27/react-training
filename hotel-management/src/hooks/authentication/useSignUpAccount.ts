import { useMutation } from '@tanstack/react-query';
import { createAccount as createAccountFn } from '@src/services/authenticationService.ts';
import toast from 'react-hot-toast';

const useSignUpAccount = () => {
  const { mutate: createAccount, isPending: isCreating } = useMutation({
    mutationFn: createAccountFn,
    onSuccess: (account) => {
      toast.success(`Create user ${account!.user!.email} successful!`);
    },
    onError: (err) => toast.error(err.message),
  });

  return { createAccount, isCreating };
};

export { useSignUpAccount };
