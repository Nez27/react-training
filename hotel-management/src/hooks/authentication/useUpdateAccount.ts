import { useMutation, useQueryClient } from '@tanstack/react-query';
import { toast } from 'react-hot-toast';

// Services
import { updateAccount as updateAccountFn } from '@service/authenticationService';

// Constants
import { UPDATE_ACCOUNT_SUCCESSFUL } from '@constant/messages';

const useUpdateAccount = () => {
  const queryClient = useQueryClient();

  const { mutate: updateAccount, isPending: isUpdating } = useMutation({
    mutationFn: updateAccountFn,
    onSuccess: (account) => {
      toast.success(UPDATE_ACCOUNT_SUCCESSFUL);
      queryClient.setQueryData(['account'], account.user);
    },
    onError: (err) => toast.error(err.message),
  });

  return { updateAccount, isUpdating };
};

export { useUpdateAccount };
