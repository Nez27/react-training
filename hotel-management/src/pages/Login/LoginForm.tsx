import { useForm } from 'react-hook-form';

// Components
import Form from '@src/components/Form';

// Styled
import { FieldInput, StyledLoginForm } from './styled';
import Button from '@src/commons/styles/Button';

// Constants
import {
  INVALID_EMAIL,
  REQUIRED_FIELD_ERROR,
} from '@src/constants/formValidateMessage';

// Types
import { ILogin } from '@src/types/common';

// Hooks
import { useLogin } from '@src/hooks/authentication/useLogin';

// Helpers
import { isValidRegex } from '@src/helpers/validators';

// Constants
import { REGEX } from '@src/constants/commons';

const LoginForm = () => {
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<ILogin>({
    defaultValues: {
      email: 'admin@nezumi.site',
      password: 'Phanhuuloi2@@1',
    }
  });
  const { login, isPending } = useLogin();

  const onSubmit = (data: ILogin) => {
    login(data);
  };

  return (
    <StyledLoginForm>
      <Form onSubmit={handleSubmit(onSubmit)}>
        <Form.Row
          label="Email:"
          direction="vertical"
          error={errors.email?.message}
        >
          <FieldInput
            type="text"
            id="email"
            {...register('email', {
              required: REQUIRED_FIELD_ERROR,
              validate: (value) =>
                isValidRegex(REGEX.EMAIL, value) || INVALID_EMAIL,
            })}
          />
        </Form.Row>
        <Form.Row
          label="Password:"
          direction="vertical"
          error={errors.password?.message}
        >
          <FieldInput
            type="password"
            id="password"
            {...register('password', {
              required: REQUIRED_FIELD_ERROR,
            })}
          />
        </Form.Row>

        <Button type="submit" disabled={isPending}>
          Login
        </Button>
      </Form>
    </StyledLoginForm>
  );
};

export default LoginForm;
