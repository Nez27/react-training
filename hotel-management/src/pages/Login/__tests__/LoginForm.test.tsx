import { useLogin } from "@hook/authentication/useLogin";
import LoginForm from "@page/Login/LoginForm";
import { fireEvent, render, waitFor } from "@testing-library/react";


// Mock the useLogin hook
jest.mock('@hook/authentication/useLogin', () => ({
  useLogin: jest.fn(),
}));

describe('LoginForm Component', () => {
  const mockLogin = jest.fn();

  beforeEach(() => {
    (useLogin as jest.Mock).mockReturnValue({
      login: mockLogin,
      isPending: false,
    });
  })

  test('Should render correctly', () => {
    const wrapper = render(<LoginForm />);
    
    expect(wrapper).toMatchSnapshot();
  });

  test('submits the form with valid data', async () => {
    const wrapper = render(<LoginForm />);

    // Submit form with valid data
    fireEvent.input(wrapper.container.querySelector('#email')!, {
      target: { value: 'admin@nezumi.site' },
    });
    fireEvent.input(wrapper.container.querySelector('#password')!, {
      target: { value: '@Ccac71b36122cd' },
    });
    fireEvent.click(wrapper.getByText('Login'));

    await waitFor(() => {
      expect(mockLogin).toHaveBeenCalledWith({
        email: 'admin@nezumi.site',
        password: '@Ccac71b36122cd',
      });
    });
  });

  test('displays error message for invalid email', async () => {
    const wrapper = render(<LoginForm />);

    // Submit form with invalid email
    fireEvent.input(wrapper.container.querySelector('#email')!, {
      target: { value: 'nezumi' },
    });
    fireEvent.click(wrapper.getByText('Login'));

    // Wait for validation error message
    await waitFor(() => {
      expect(wrapper.getByText('Invalid email address')).toBeInTheDocument();
    });
  });

  // Add more test cases as needed
});
