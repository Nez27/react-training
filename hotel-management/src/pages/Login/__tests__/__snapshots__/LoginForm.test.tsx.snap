// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm Component Should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-jrsKJM kJxxfd"
      >
        <form
          class="sc-gtsqUy jMGskL"
        >
          <div
            class="sc-hKFymg jzIDWp"
            direction="vertical"
          >
            <label
              class="sc-eCAqax dnXnhF"
            >
              Email:
            </label>
            <div>
              <input
                class="sc-kEqXeH jhZfUU"
                id="email"
                name="email"
                type="text"
              />
            </div>
          </div>
          <div
            class="sc-hKFymg jzIDWp"
            direction="vertical"
          >
            <label
              class="sc-eCAqax dnXnhF"
            >
              Password:
            </label>
            <div>
              <input
                class="sc-kEqXeH jhZfUU"
                id="password"
                name="password"
                type="password"
              />
            </div>
          </div>
          <button
            class="sc-bdnyFh bEtbaY"
            type="submit"
            variations="primary"
          >
            Login
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-jrsKJM kJxxfd"
    >
      <form
        class="sc-gtsqUy jMGskL"
      >
        <div
          class="sc-hKFymg jzIDWp"
          direction="vertical"
        >
          <label
            class="sc-eCAqax dnXnhF"
          >
            Email:
          </label>
          <div>
            <input
              class="sc-kEqXeH jhZfUU"
              id="email"
              name="email"
              type="text"
            />
          </div>
        </div>
        <div
          class="sc-hKFymg jzIDWp"
          direction="vertical"
        >
          <label
            class="sc-eCAqax dnXnhF"
          >
            Password:
          </label>
          <div>
            <input
              class="sc-kEqXeH jhZfUU"
              id="password"
              name="password"
              type="password"
            />
          </div>
        </div>
        <button
          class="sc-bdnyFh bEtbaY"
          type="submit"
          variations="primary"
        >
          Login
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
